### PR TITLE
fix: Correct match-based-id for lockfile-only SCA findings

### PR DESCRIPTION
**Pull request by mmcqd**
https://www.notion.so/semgrep/Fixing-SSC-match-based-id-and-triage-state-d4676bfa62a2439eb026ac22b3d6c119

Given a lockfile with two instances of a library, say `foo@1.0.0` and `foo@1.0.1`, if both are matched by a lockfile-only SCA rule, semgrep will produce two findings, but these findings will incorrectly have identical `match_based_id`s, so Semgrep App considers them identical findings. This happened because lockfile-only findings were created without the use of the `RuleMatches` class, which handles updating the `match_based_id` of each finding. We actually can't use `RuleMatches`, because it ends up changing the whole `match_based_id`, but we only want to change the final index, to avoid duplicating findings in the database, so we re-implement this functionality manually.

In the case where the lockfile contains two instances of a library at the _same_ version, then we will still produce two separate findings with different `match_based_id`s. This is technically incorrect, but all of our code is built assuming this is possible, so it can be corrected in future PRs.

Due basically to luck, for a given set of lockfile-only findings from a single rule, the Issue with `match_based_id` ending with index 0 in the database will be the same before and after this change. This is because before upload to the database these findings are put into a dictionary keyed on their `match_based_id`, and the findings are processed in reverse order, so the dependency that appeared _first_ in a lockfile gets put into this dictionary _last_. Before the change, when these findings all had the same `match_based_id`, the first matched dependency in the lockfile becomes the only thing in the database. After the change, the first thing in the lockfile still gets index 0, since it appears first. This means the single issue that gets selected as the representative of the vuln group for these issues won't change, since we select the oldest issue, and the one ending in index 0 existed previously, so it's the oldest.

### CODE REVIEW

File: cli/src/semgrep/dependency_aware_rule.py [hunk 1]
```diff
@@ -122,8 +123,8 @@ def generate_unreachable_sca_findings(
                         )
                     },
                 )
-                non_reachable_matches.append(match)
-    return non_reachable_matches, dep_rule_errors
+                non_reachable_matches.add(match)
+    return sorted(non_reachable_matches), dep_rule_errors
```
Comment (gautambhat) [inline cli/src/semgrep/dependency_aware_rule.py]: What is the reason for sorting it?

File: cli/tests/default/unit/test_match_hashes.py [hunk 2]
```diff
@@ -50,6 +58,28 @@ def double_eqeq_rule() -> Rule:
...
+            """
+        rules:
+        - id: rule_id
+          r2c-internal-project-depends-on:
+              package: "foo"
+              version: ">=1.0.0"
+              namespace: "pypi"
+          languages: [python]
+          severity: INFO
+          message: bad
```
Comment (gautambhat) [inline cli/tests/default/unit/test_match_hashes.py:75]: love it

File: cli/tests/default/unit/test_match_hashes.py [hunk 3]
```diff
@@ -217,3 +293,22 @@ def test_same_code_hash_for_previous_scan_finding(mocker, foo_contents):
...
+        foo == 3.0.0
+        """
+    ).lstrip()
+    mocker.patch.object(Path, "open", mocker.mock_open(read_data=lockfile_contents))
+    match1 = get_lockfile_only_rule_match(start_line=1, end_line=1)
+    match2 = get_lockfile_only_rule_match(start_line=2, end_line=2)
+    matches = RuleMatches(lockfile_only_rule)
+    matches.update([match1, match2])
+    matches = list(sorted(matches))
+    assert matches[0].match_based_id != matches[1].match_based_id
```
Comment (gautambhat) [inline cli/tests/default/unit/test_match_hashes.py:314]: Just wanna verify that this fails without your changes 😬

**Review by gautambhat — APPROVED** [review]
lgtm

File: cli/src/semgrep/dependency_aware_rule.py (hunk 1 above)
Comment (mmcqd) [reply cli/src/semgrep/dependency_aware_rule.py]: Changed this

**Pull request merged**